datafusion/core/src/dataframe.rs

```diff
@@ -1515,4 +1515,84 @@ mod tests {
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn partition_aware_union() -> Result<()> {
+        let left = test_table().await?.select_columns(&["c1", "c2"])?;
+        let right = test_table_with_name("c2")
+            .await?
+            .select_columns(&["c1", "c3"])?
+            .with_column_renamed("c2.c1", "c2_c1")?;
+
+        let left_rows = left.collect().await?;
+        let right_rows = right.collect().await?;
+        let join1 =
```

Comment:
Could you possibly add some comments here about what this test is verifying? It seems like perhaps it is verifying that when the joins are on the same key the partitioning is  the same and thus union can be done without bringing everything to a single stream?

Reply:
Sure, this UT verifies the union is partition aware, by checking the output partitioning count is the default partition count(not the combinations of the input) and also the result is correct. I will add more comments here, and I will also add more code here to verify the output partition is same with the union's inputs.